### DISCUSSION TITLE
List.RemoveAt() should support using Index, and maybe even Range?

### DISCUSSION THREAD

**Discussion by jklw10** [dotnet/csharplang#6941] [Language Ideas]
at the moment doing
``List.RemoveAt(^1)``
is invalid.
also, ``List.RemoveRange()`` should support System.Range.

Currently only integers are supported for both methods.

**Comment by spydacarnage** [comment]
This isn't a language change - it should be addressed in the [runtime](https://github.com/dotnet/runtime) repo.

**Comment by DavidArno** [comment]
```cs
public static class ListExtensions 
{
    public static void RemoveAt<T>(this List<T> list, Index index) 
        => list.RemoveAt(IndexToPosition(index, list.Count));
    
    public static void RemoveRange<T>(this List<T> list, Range range)
    {
        var count = list.Count;
        var startPosition = IndexToPosition(range.Start, count);
        list.RemoveRange(startPosition, IndexToPosition(range.End, count) - startPosition);
    }
    
    private static int IndexToPosition(Index index, int count) 
        => index.Value is var i && index.IsFromEnd ? count - i : i;
}
```

**Reply by CyrusNajmabadi** [reply]
@DavidArno IndexToPosition can be replaced with https://learn.microsoft.com/en-us/dotnet/api/system.index.getoffset?view=net-7.0#system-index-getoffset(system-int32)

**Reply by DavidArno** [reply]
Nice. Thanks for the tip, Cyrus. So the code can be reduced to just this:

```cs
public static class ListExtensions 
{
    public static void RemoveAt<T>(this List<T> list, Index index) 
        => list.RemoveAt(index.GetOffset(list.Count));
    
    public static void RemoveRange<T>(this List<T> list, Range range)
    {
        var count = list.Count;
        var startPosition = range.Start.GetOffset(count);
        list.RemoveRange(startPosition, range.End.GetOffset(count) - startPosition);
    } 
}
```

**Reply by jklw10** [reply]
i must say i completely forgot about extensions, awesome stuff.  I've been told this is apparently runtime's domain of change, any rules for how i should proceed with the whole thing? copy the post and link discussion?

**Reply by 333fred** [reply]
This is the process: https://github.com/dotnet/runtime/blob/main/docs/project/api-review-process.md#api-review-process
